### ISSUE TITLE
XmlSerializer serializes Nullable types with `null` as node value

### ISSUE THREAD

**Issue by danielmarbach** [Particular/NServiceBus#3122]
## Who's affected
Customers doing integration with other systems over messaging which are not NServiceBus endpoints and using the `XmlSerializer` with messages containing `Nullable` properties.

## Symptoms
Given the following message definition:
```
    public class MessageWithNullable : IMessage
    {
        public string FirstName { get; set; }

        public string LastName { get; set; }

        public string EmailAddress { get; set; }

        public DateTime? BirthDate { get; set; } //Nullable DateTime property
    }
```

we serialize it to

```
<?xml version="1.0"?>
<MessageWithNullable xmlns="http://tempuri.net/NServiceBus.Serializers.XML.Test" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
   <FirstName>FirstName</FirstName>
   <LastName>LastName</LastName>
   <EmailAddress>EmailAddress</EmailAddress>
   <BirthDate>null</BirthDate>
</MessageWithNullable>
```

We should serialize it to

```
<?xml version="1.0"?>
<MessageWithNullable xmlns="http://tempuri.net/NServiceBus.Serializers.XML.Test" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
   <FirstName>FirstName</FirstName>
   <LastName>LastName</LastName>
   <EmailAddress>EmailAddress</EmailAddress>
   <BirthDate />
</MessageWithNullable>
```

or even better

```
<?xml version="1.0"?>
<MessageWithNullable xmlns="http://tempuri.net/NServiceBus.Serializers.XML.Test" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
   <FirstName>FirstName</FirstName>
   <LastName>LastName</LastName>
   <EmailAddress>EmailAddress</EmailAddress>
   <BirthDate xsi:nil="true" />
</MessageWithNullable>
```

http://www.w3.org/TR/xmlschema-1/#xsi_nil
http://stackoverflow.com/questions/774192/what-is-the-correct-way-to-represent-null-xml-elements

https://groups.google.com/forum/#!searchin/particularsoftware/NSB$20XmlSerializer$20-$20Serialization$20of$20Nullable$20types/particularsoftware/Pl4StwS44-w/5oVdh4vRBAAJ

**Comment by danielmarbach** [comment]
Also related to https://github.com/Particular/NServiceBus/issues/2706

See 

https://github.com/Particular/NServiceBus/blob/master/src/NServiceBus.Core/Serializers/XML/Serializer.cs#L370

and

https://github.com/Particular/NServiceBus/blob/master/src/NServiceBus.Core/Serializers/XML/Serializer.cs#L411

**Comment by danielmarbach** [comment]
I think we need to fix #2706 as well. In my point of view the correct way to serialize #2706 would be

```
<?xml version=""1.0"" ?>
<MessageWithNullableArray xmlns:xsi=""http://www.w3.org/2001/XMLSchema-instance"" xmlns:xsd=""http://www.w3.org/2001/XMLSchema"" xmlns=""http://tempuri.net/NServiceBus.Serializers.XML.Test"">
   <SagaId>00000000-0000-0000-0000-000000000000</SagaId>
   <SomeInts>
      <NullableOfInt32 />
   </SomeInts>
</MessageWithNullableArray>
````

**Comment by danielmarbach** [comment]
The story gets even worse. Fix #2706 makes now this when you have the following array

```
new int?[] { null, 1}
```

```
<?xml version="1.0"?>
<MessageWithNullableArray xmlns="http://tempuri.net/NServiceBus.Serializers.XML.Test" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
   <SagaId>00000000-0000-0000-0000-000000000000</SagaId>
   <SomeInts>
      <NullableOfInt32> null</NullableOfInt32>
      <Int32>1</Int32>
   </SomeInts>
</MessageWithNullableArray>
```

In my opinion the outcome should be

```
<?xml version="1.0"?>
<MessageWithNullableArray xmlns="http://tempuri.net/NServiceBus.Serializers.XML.Test" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
   <SagaId>00000000-0000-0000-0000-000000000000</SagaId>
   <SomeInts>
      <NullableOfInt32 />
      <NullableOfInt32>1</Int32>
   </SomeInts>
</MessageWithNullableArray>
```

**Comment by danielmarbach** [comment]
The problem is how we determine content of `IEnumerable`

https://github.com/Particular/NServiceBus/blob/master/src/NServiceBus.Core/Serializers/XML/Serializer.cs#L478

I see the value in serializing objects with their serialization friendly names when you have list with complex types which can be of different reference types like for example

```
var animals = new List<Animal> { new Dog(), new Cat() }
```

but we cannot determine wether an enumerable contains simple types during the iteration of the enumerable. We have to change the logic to determine on the level of the enumerable whether it is an enumerable of simple types or complex types. When we determined it contains simple types we have to treat all elements in that enumerable with the same type.

**Comment by danielmarbach** [comment]
@andreasohlund @johnsimons @distantcam @SeanFeldman please chime in since you have been involved in #2706

**Comment by danielmarbach** [comment]
The even more robust representation of the above scenarios would be:

```
<?xml version="1.0"?>
<MessageWithNullable xmlns="http://tempuri.net/NServiceBus.Serializers.XML.Test" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
   <FirstName>FirstName</FirstName>
   <LastName>LastName</LastName>
   <EmailAddress>EmailAddress</EmailAddress>
   <BirthDate xsi:nil="true" />
</MessageWithNullable>
```

```
<?xml version="1.0"?>
<MessageWithNullableArray xmlns="http://tempuri.net/NServiceBus.Serializers.XML.Test" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
   <SagaId>00000000-0000-0000-0000-000000000000</SagaId>
   <SomeInts>
      <NullableOfInt32 xsi:nil="true" />
      <NullableOfInt32>1</Int32>
   </SomeInts>
</MessageWithNullableArray>
```

http://www.w3.org/TR/xmlschema-1/#xsi_nil
http://stackoverflow.com/questions/774192/what-is-the-correct-way-to-represent-null-xml-elements

**Comment by ramonsmits** [comment]
Null should be serialized as `< BirthDate xsi:nil="true"/>` right? Having `<BirthDate />` is an improvement.

That collection serialization is horrible :-(

**Comment by danielmarbach** [comment]
@ramonsmits https://github.com/Particular/NServiceBus/issues/3122#issuecomment-159950628

**Comment by ramonsmits** [comment]
@danielmarbach Sorry, missed that. Maybe update the description?

**Comment by danielmarbach** [comment]
@ramonsmits done

**Comment by SeanFeldman** [comment]
@danielmarbach so should it be as in description right now
```xml
<BirthDate />
```
or (as in [comment](https://github.com/Particular/NServiceBus/issues/3122#issuecomment-159950628))
```xml
<BirthDate xsi:nil="true" />
```

**Comment by danielmarbach** [comment]
@SeanFeldman I updated the issue card above. It shows now both approaches

**Comment by petersgiles** [comment]
Why do we have our own implementation of XmlSerializer? I'm guessing its probably because nothing was good enough back in the day when it was first written, but AFAIK nowadays the native one in .Net is pretty good. Would it be possible to deprecate our one and swap in the regular one. That way it will always behave as Devs expect.

**Comment by danielmarbach** [comment]
@petersgiles At that time our serializer was faster and more capable. I think it still support more types out of the box than the standard serializer. With that added value comes the prize we pay to keep it bug free. Swapping serializer is hard because we need to make it backward compat on the wire... Maybe in the long term roadmap vor vNextNext?

**Comment by petersgiles** [comment]
Ok it's definitely worth a conversation https://github.com/Particular/NServiceBus/issues/3128

**Comment by johnsimons** [comment]
@danielmarbach your suggestion to use the `xsi:nil="true"` seems to be the way forward IMO.
But this is going to be really hard to provide forward and backwards compatibility.

**Comment by WilliamBZA** [comment]
An internal fix for this is relatively trivial. I have a concern around compatibility and backporting though:

This is a message format change, which is going to be a problem when dealing with endpoints running different versions of the serializer/deserializer.  We can quite easily make the fix serialize the new (nil) format, while making it able to deserialize both the old as well as the new format.

This will mean that endpoints running the new version of the schema will be able to read messages coming from old endpoints. Old endpoints won't be able to deserialize the new message format though and the messages will end up in the error queue.

This becomes even messier when dealing with non-NServiceBus endpoints that expect messages to be in the old format and have actually worked around the incorrect XML. There is no guarantee that clients will even be able to update those systems.

TL;DR: Quite a big breaking change.

Due to this, I propose the following:
- We do this as a v6 (or vNext) feature only
- We mark the custom serializer as obsolete scheduled for removal in v7 (to encourage users to use the standard .Net serializer and prevent issues like this in the future)
- We DON'T backport the fix for this
- We document how to migrate all endpoints to the new schema (perhaps provide a tool to update the messages in the error queue to the new format and automatically resend them)

**Comment by andreasohlund** [comment]
>We do this as a v6 (or vNext) feature only

Since this breaks wire compatiblity (the holy grail of NServiceBus) I'd say :-1:


Another option would be to create a "XmlSerializerV2" that has the behavior or make it opt-in?

**Comment by WilliamBZA** [comment]
@andreasohlund that could work. If it becomes opt-in, does that affect the likelihood of success of backporting?

We'll still have breaking contracts in old endpoints receiving new format messages until all endpoints are updated.

**Comment by danielmarbach** [comment]
@udidahan does wire compatibility apply for NSB to NSB only or also for third parties?

**Comment by andreasohlund** [comment]
I'd say NSB to NSB is must have while third party integrations are best
effort?

**Comment by WilliamBZA** [comment]
This just in: The existing deserializer can deserialize both the new xsi:nil format as well as the previous null format without any modifications. This made the change much simpler with a lot less risk.

To summarize:

Old NSB to new NSB endpoints will integrate just fine.
New NSB to old NSB endpoints will integrate just fine.

It is also highly likely (but definitely not guaranteed) that new NSB will be able to integrate with 3rd parties expecting the old format.

**Comment by SimonCropp** [comment]
@williambza did u verify that with older versions as well?

**Comment by danielmarbach** [comment]
@WilliamBZA but the idea is that the patched version of the serializer produces the new format, correct?

**Comment by WilliamBZA** [comment]
@SimonCropp Verified in:
- current develop
- current master
- 5.2.12
- 5.0.0
- 4.7.11
- 4.0.1

@danielmarbach Correct. What I'm saying is that the new format turns out to be backward compatible so older endpoints won't break when a new endpoint starts sending them messages in the new format.

**Comment by danielmarbach** [comment]
> It is also highly likely (but definitely not guaranteed) that new NSB will be able to integrate with 3rd parties expecting the old format.

I didn't understand the above statement. Since we are sending the new format how is that highlz likely that new NSB will be able to integrate with 3rd parties?

**Comment by danielmarbach** [comment]
Can you also check 3.3.x?

**Comment by WilliamBZA** [comment]
@danielmarbach It's a guess at this point, I'm basing it on the fact that our deserializer didn't need any tweaks to be able to work with both formats. So if 3rd parties have similar code to ours, theirs might also just work.

They can't be using the built-in .NET deserializer, as that doesn't work for the old format - only the new format, so there has to be some custom code in play - hence my guess that their custom code would be similar to ours.

Busy checking 3.3.x now...

**Comment by andreasohlund** [comment]
>Busy checking 3.3.x now...

Are you using the test suite that @hmemcpy created? https://github.com/Particular/ProductionTests/tree/master/src/SerializerCompatibilityTests

**Comment by SimonCropp** [comment]
@WilliamBZA thanks for the version depth check

**Comment by WilliamBZA** [comment]
@andreasohlund I wasn't, I am now.

@SimonCropp @danielmarbach Confirmed through the SerializerCompatibilityTests the new version of the format can be successfully deserialized from version 3.3.18 updwards.

**Comment by andreasohlund** [comment]
>Confirmed through the SerializerCompatibilityTests the new version of the format can be successfully deserialized from version 3.3.18 updwards.

:dancer:

**Comment by WilliamBZA** [comment]
Do we have a consensus on backporting this yet? I still question the value given that backward compatibility seems to be unaffected for our code, but could still be affected for 3rd party code.

**Comment by andreasohlund** [comment]
Since we haven't heard of anyone beeing effected I'd say to not backport for now?

**Comment by danielmarbach** [comment]
@andreasohlund There is a google case linked. So apparently someone is
affected ;)

**Comment by andreasohlund** [comment]
Dang, thought you found it during testing @danielmarbach . In that case we need to get in touch with the customer and see what the current state it.

FYI https://github.com/Particular/PlatformDevelopment/blob/master/Guidelines/Patching.md

**Comment by WilliamBZA** [comment]
@danielmarbach I can't see a linked google case anywhere? There is a support case that was closed a while back, but that's the only one I can see linked to this.

**Comment by danielmarbach** [comment]
@andreasohlund @WilliamBZA https://groups.google.com/forum/#!searchin/particularsoftware/NSB$20XmlSerializer$20-$20Serialization$20of$20Nullable$20types/particularsoftware/Pl4StwS44-w/5oVdh4vRBAAJ

**Comment by hmemcpy** [comment]
> Confirmed through the SerializerCompatibilityTests the new version of the format can be successfully deserialized from version 3.3.18 updwards.

:dancers: :confetti_ball: 

cc-ing @tmasternak just so he gets a notification :)

**Comment by SimonCropp** [comment]
-1 to backporting

**Comment by danielmarbach** [comment]
Why?

**Comment by SimonCropp** [comment]
Not every bugs needs to be backported. There is an ROI to consider. U could also argue this is not a bug. It is just how it was serialized.

In my view the rarity of this occuring, the very few people effected, and the "is it really" a bug means the value of backporting is not worth the pain and effort

**Comment by danielmarbach** [comment]
Well from an integration perspective when it comes to do interop not
adhering a standard like XML is for me a big argument to do it. BUT since
there is a workaround by using

http://docs.particular.net/nservicebus/serialization/xml#raw-xml

I'm ok with not backporting it. Should we instead provide a sample using
raw-xml for that specific scenario?

**Comment by WilliamBZA** [comment]
The linked google issue is on version 5.2.9 so perhaps we should backport to version 5.* since there is an actual issue there, but leave the earlier versions?

At any rate, the PR is ready for review.

**Comment by danielmarbach** [comment]
@WilliamBZA @andreasohlund since this is breaking third parties I don't think we should bring this into V5. Thoughts?

**Comment by andreasohlund** [comment]
Sounds resonable to me
